semantic.py

```diff
@@ -29,24 +44,236 @@ def analyze(self):
         for i in self.tree.children:
             print(i.pretty())
 
+    # TODO order these checks in a better order
+
     # Check that the data has a start symbol
     def is_valid_tree(self, tree):
-        return tree.data == 'start'
+        if not type(tree) is self.treetype:
+            return False
+
+        if tree.data != 'start':
+            return False
+        for i in range(len(tree.children[:-1])):
+            if tree.children[i].data == 'id':
+                if tree.children[i+1].data != 'rhs':
+                    return False
+            if tree.children[i].data == 'rhs':
+                if tree.children[i-1].data != 'id':
+                    return False
+        if tree.children[-1].data != 'compose':
+            return False
+        return True
+
+    def is_valid_dynamic(self, tree):
+        if not type(tree) is self.treetype:
+            return False
+        if tree.data != 'dynamic':
+            return False
+
+        item = tree.children[0].data
+        if item == 'inlinedynamic':
+            d = tree.children[0].children[0].lower()
+            if d not in self.valid_levels:
+                return False
+        elif item == 'id':
+            return Semantic.is_valid_identifier(self, tree)
+        else:
+            return False
+
+        return True
+
+    def is_valid_inlinedynamic(self, tree):
+        if not type(tree) is self.treetype:
+            return False
+        if tree.data != 'inlinedynamic':
+            return False
+
+        d = tree.children[0].lower()
+        if d not in self.valid_levels:
+            return False
+
+        return True
+
+    def is_valid_note(self, tree):
+        if not type(tree) is self.treetype:
+            return False
+
+        if len(tree.children[0].children) != 2:
+            return False
+
+        if tree.children[0].children[0].data != 'division':
+            return False
+        else:
+            n = tree.children[0].children[1].data
+            if not ( n == 'notename' or n == 'chord' or n == 'tuple' or n == 'id' or n == 'REST'):
```

Comment:
You have quite a few lines like this, where you check
```
if not (n == this || n == that || n = anything)
```
while you should actually be saying 
```
if n not in self.valid_data
```

This will make the code a little more readable, and also easier to change. Additionally it will have less 'magic strings' and 'magic numbers.